### PR TITLE
ENH: Pattern computation for linear models

### PR DESCRIPTION
**Pull request by trachelr**
This is a simple function that fit a linear model onto some data (X, y), computes the corresponding patterns and create an evoked object with it.
Visualization is much more simple now...

### CODE REVIEW

File: mne/decoding/classifier.py [hunk 1]
```diff
@@ -453,3 +453,23 @@ def transform(self, epochs_data, y=None):
                                      picks=self.picks, n_jobs=self.n_jobs,
                                      copy=False, verbose=False)
         return epochs_data
+
+
+def compute_patterns(epochs, linear_model):
```
Comment (mainakjas) [inline mne/decoding/classifier.py]: so this PR and https://github.com/mne-tools/mne-python/pull/2282 are achieving the same purpose? or are they somehow different?

File: mne/decoding/classifier.py [hunk 2]
```diff
@@ -453,3 +453,23 @@ def transform(self, epochs_data, y=None):
...
+    X = sc.fit_transform(X)
+    # fit the linear model
+    linear_model.fit(X, labels)
+    # predict the train data
+    y_pred = linear_model.predict(X)
+    # computes the patterns and reshape it
+    patterns = np.dot(y_pred, X).reshape(len(epochs.ch_names), 
+                                         len(epochs.times))
+    evoked.data = patterns
+    return evoked
```
Comment (mainakjas) [inline mne/decoding/classifier.py]: new line missing

File: mne/decoding/classifier.py (hunk 1 above)
Comment (trachelr) [inline mne/decoding/classifier.py]: They are quite different, this code is related to the paper by Haufe et al, 2014 (neuroimage).
It computes patterns from a linear model (e.g. a RidgeClassifier) and put it into an evoked object which is more convenient for visualisation.

Comment (trachelr) [inline mne/decoding/classifier.py]: I'm making an example to explain how it works

File: examples/decoding/plot_decoding_linear_model_patterns.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+"""
+====================================================================
+Decoding in sensor space data using the linear models 
+====================================================================
+
+Decoding applied to MEG data in sensor space using a linear classifier
+Here the model coefiscient are interpreted using [1].
```
Comment (agramfort) [inline examples/decoding/plot_decoding_linear_model_patterns.py]: coefiscient :)

File: examples/decoding/plot_decoding_linear_model_patterns.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+raw.info['bads'] = ['MEG 2443']  # set bad channels
+picks = mne.pick_types(raw.info, meg='grad', eeg=False, stim=False, eog=False,
+                       exclude='bads')
+
+# Read epochs
+epochs = mne.Epochs(raw, events, event_id, tmin, tmax, proj=True,
+                    picks=picks, baseline=None, preload=True)
+
+labels = epochs.events[:, -1]
```
Comment (agramfort) [inline examples/decoding/plot_decoding_linear_model_patterns.py]: labels is not used

File: mne/decoding/classifier.py [hunk 5]
```diff
@@ -453,3 +453,24 @@ def transform(self, epochs_data, y=None):
...
+    sc = StandardScaler()
+    X = sc.fit_transform(X)
+    # fit the linear model
+    linear_model.fit(X, labels)
+    # predict the train data
+    y_pred = linear_model.predict(X)
+    # computes the patterns and reshape it
+    patterns = np.dot(y_pred, X).reshape(len(epochs.ch_names), 
+                                         len(epochs.times))
+    evoked.data = patterns
```
Comment (agramfort) [inline mne/decoding/classifier.py]: that's not clean. Can you use EvokedArray?

**Comment by agramfort** [comment]
also you'll need to make travis happy

File: mne/decoding/classifier.py [hunk 6]
```diff
@@ -453,3 +453,24 @@ def transform(self, epochs_data, y=None):
...
+    # normalize the data
+    sc = StandardScaler()
+    X = sc.fit_transform(X)
+    # fit the linear model
+    linear_model.fit(X, labels)
+    # predict the train data
+    y_pred = linear_model.predict(X)
+    # computes the patterns and reshape it
+    patterns = np.dot(y_pred, X).reshape(len(epochs.ch_names), 
+                                         len(epochs.times))
```
Comment (wmvanvliet) [inline mne/decoding/classifier.py]: Which formula in Haufe 2014 is this implementing? Is this equivalent to formula 6?

File: mne/decoding/classifier.py (hunk 1 above)
Comment (wmvanvliet) [inline mne/decoding/classifier.py]: This can be a nice unit test. First, fit a CSP and compute weights and patterns in the traditional way (inverse(W).T). Then compare with the patterns produced by this method. If the code is correct, they should be the same.

Comment (trachelr) [inline mne/decoding/classifier.py]: Thanks for the suggestion. I'll do it

File: mne/decoding/classifier.py (hunk 6 above)
Comment (trachelr) [inline mne/decoding/classifier.py]: Yes it is a simplification of formula 6, case where K=1 see formula 7
In case of K > 1, we should use an OLS approach formula 8,9

File: examples/decoding/plot_decoding_linear_model_patterns.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+# computes patterns estimated with a linear SVM classifier
+linsvc = svm.LinearSVC()
+linsvc_patterns = compute_patterns(epochs, linsvc)
+linsvc_patterns.plot_topomap()
+
+# computes patterns estimated with a logistic regression classifier
+logreg = lm.LogisticRegression(penalty='l1')
+logreg_patterns = compute_patterns(epochs, logreg)
+logreg_patterns.plot_topomap()
```
Comment (kingjr) [inline examples/decoding/plot_decoding_linear_model_patterns.py]: I think the example only needs one linear classifier, and perhaps one linear regresser to make the point.

Comment (agramfort) [inline examples/decoding/plot_decoding_linear_model_patterns.py]: +1

Comment (trachelr) [inline examples/decoding/plot_decoding_linear_model_patterns.py]: Right, maybe a comparison with a t-test value would make the link with classic univariate analysis...

File: mne/decoding/classifier.py [hunk 8]
```diff
@@ -453,3 +453,23 @@ def transform(self, epochs_data, y=None):
...
+
+
+def compute_patterns(epochs, linear_model):
+    from sklearn.preprocessing import StandardScaler
+    labels = epochs.events[:, -1]
+    # concatenate channel x time data
+    X = epochs.get_data().reshape(len(epochs), -1)
+    # normalize the data
+    sc = StandardScaler()
+    X = sc.fit_transform(X)
```
Comment (kingjr) [inline mne/decoding/classifier.py]: I think you should keep the fitting and transform stage separated. If one forgets to fit using your method, he/she could still want to look at the patterns.

File: mne/decoding/classifier.py (hunk 6 above)
Comment (wmvanvliet) [inline mne/decoding/classifier.py]: Formula 7 states:

    A = dot(cov(X), W) = cov(X, y_pred)

In your code, a simple dot multiplication is performed. I'm assuming this boils down to the emperical estimate of the covariance. It would be nice to support some [shrinkage](http://scikit-learn.org/stable/modules/covariance.html) in there as well.

Why not do the OLS in all cases? Formula 6 is just the special K=1 case. When treating the problem as a regression problem you gain a lot of benefits (as pointed out in the paper), such as straightforward ways to incorporate shrinkage and other priors. This approach should gracefully degrade to the computation you perform now when K=1.

Comment (agramfort) [inline mne/decoding/classifier.py]: the regularization by shrinkage should appear in y_pred. You don't need
shrinkage in cov for patterns AFAIK

**Comment by wmvanvliet** [comment]
It would be nice if the function could also be passed a pre-fitted model, in which case `linear_model.fit()` would not be called.

From the scikit-learn manual:

> As a result the existence of parameters with trailing _ is used to check if the estimator has been fitted.

**Comment by wmvanvliet** [comment]
I would for example love to use this function to do the following:

- Fit classifier to my data
- Verify the fitting is proper by using `compute_patterns`
- Start using the classifier for fun and profit

File: examples/decoding/plot_decoding_linear_model_patterns.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+# Read epochs
+epochs = mne.Epochs(raw, events, event_id, tmin, tmax, proj=True,
+                    picks=picks, baseline=None, preload=True)
+
+import sklearn.linear_model as lm
+import sklearn.svm as svm
+from mne.decoding.classifier import compute_patterns
+
+# computes patterns estimated with a ridge classifier
+ridge  = lm.RidgeClassifier()
```
Comment (kingjr) [inline examples/decoding/plot_decoding_linear_model_patterns.py]: two spaces

File: mne/decoding/classifier.py [hunk 10]
```diff
@@ -453,3 +453,23 @@ def transform(self, epochs_data, y=None):
...
+    # fit the linear model
+    linear_model.fit(X, labels)
+    # predict the train data
+    y_pred = linear_model.predict(X)
+    # computes the patterns and reshape it
+    patterns = np.dot(y_pred, X).reshape(len(epochs.ch_names), 
+                                         len(epochs.times))
+    # return an Evoked object
+    return mne.EvokedArray(patterns, epochs.info, tmin = epochs.tmin)
```
Comment (kingjr) [inline mne/decoding/classifier.py]: no space before and after the `=` in this case

File: mne/decoding/classifier.py (hunk 6 above)
Comment (trachelr) [inline mne/decoding/classifier.py]: +1 in case K=1 cov and dot are equivalent. we decided to do this simple implementation first...
I'm going to include the general case while fixing the decoding API (see issue #2290)

**Comment by agramfort** [comment]
what's the status of this? how does it relate to CSP patterns?

**Comment by wmvanvliet** [comment]
why was this closed? I was excited to see this feature!

**Comment by agramfort** [comment]
it's a dupe with the other PR creating the LinearModel class that will come with plot_patterns method.

**Comment by trachelr** [comment]
Yep @agramfort is right, sorry I should have said that before closing the PR.
@wmvanvliet have a look at examples/decoding/plot_linear_model_patterns.py in PR #2292
It nicely show different patterns/filters on MEG vs EEG data.
